apps/builder/app/builder/features/settings-panel/use-settings-logic.ts

```diff
@@ -0,0 +1,45 @@
+import type { Instance } from "@webstudio-is/project-build";
+import store from "immerhin";
+import { type KeyboardEventHandler, useRef } from "react";
+import { instancesStore } from "~/shared/nano-states";
+
+type SettingUpdate = { label?: string };
+
+export const useSettingsLogic = ({
+  selectedInstance,
+}: {
+  selectedInstance: Instance;
+}) => {
+  const changes = useRef<SettingUpdate>({});
+
+  const set = (
+    setting: keyof SettingUpdate,
+    value: SettingUpdate[keyof SettingUpdate]
```

Comment:
Now value is not collocated with key.

I would not prematurely complicate this code. It would even more clear to have all logic inside settings panel component.

Reply:
Yeah, its a bit premature, but I think its not terrible and we will need to split it for sure

Reply:
> Now value is not collocated with key.

still getting typechecked in the sense you can set a wrong value with a wrong key

Reply:
oh nooo, since its optional an object, it now allows to set unexisting value

Reply:
ah no, all good

Reply:
I mean if the other key will have number value will be `string | number` on all keys. So you loose type safety here. You can use generic or just use named functions like setLabel for simplicity. This is not generic logic to keep it abstracted.

Reply:
true, didn't think further